src/v/cloud_storage/remote_partition.cc

```diff
@@ -271,6 +271,49 @@ class partition_record_batch_reader_impl final
                     co_await set_end_of_stream();
                     co_return storage_t{};
                 }
+                auto reader_delta = _reader->current_delta();
+                if (
+                  !_ot_state->empty()
+                  && _ot_state->last_delta() > reader_delta) {
+                    // It's not safe to call 'read_sone' with the current
+                    // offset translator state becuase delta offset of the
+                    // current reader is below last delta registred by the
+                    // offset translator. The offset translator contains data
+                    // from the previous segment and there is an inconsistency
+                    // between them.
+                    //
+                    // If the reader never produced any data we can simply reset
+                    // the offset translator state and continue. Otherwise, we
+                    // need to stop producing.
+                    //
+                    // This trick should guarantee us forward progress. If the
+                    // reader will stop right away before it will be able to
+                    // produce any batches (in the first branch) that belong to
+                    // the current segment the next fetch request will likely
+                    // have start_offset that corresponds to the previous
+                    // segment. In this case the reader will have to skip all
+                    // previously seen batches and _first_produced_offset will
+                    // be set to default value. This will allow reader to reset
+                    // the ot_state and move to the current segment in the
+                    // second branch. This is safe because the ot_state won't
+                    // have any information about the previous segment.
+                    vlog(
+                      _ctxlog.info,
+                      "Detected inconsistency. Reader config: {}, delta offset "
+                      "of the current reader: {}, delta offset of the offset "
+                      "translator: {}, first offset produced by this reader: "
+                      "{}",
+                      _reader->config(),
+                      reader_delta,
+                      _ot_state->last_delta(),
+                      _first_produced_offset);
+                    if (_first_produced_offset != model::offset{}) {
+                        co_await set_end_of_stream();
+                        co_return storage_t{};
```

Comment:
This is clever, though I wonder if it's true to the Kafka protocol -- is it always "allowed" to return empty? I suppose it is in streaming cases since it's expected consumers will just keep trying, but just want to be sure.

I wonder if it makes sense to return some status code to the Kafka layer to instantiate a new reader?

Reply:
It only takes this path if something was already returned to the client. Otherwise it takes the other path which just resets the ot_state and continues. So it won't return empty result to the client. If it will the code will loop forever because the client will resubmit its fetch request and will get another empty result in response.